### PR TITLE
HMS integration with Discord Roles

### PR DESCRIPTION
**Pull request by AaronJackson**
As discussed in a couple of recent members meetings there are plans to migrate from Slack to Discord as it provides better management of roles and doesn't restrict message history on free accounts. By default, people who join the Discord will be placed in a public channel until the appropriate roles are defined (e.g. Current Member).

This pull request adds the following:

- A new field in the user profile called `discordUserId`, which a user set for the purposes of role synchronisation. I've included the doctrine migration script for this.
- An event listener for `UserRemovedFromRole` and `UserAddedToRole` which makes the appropriate changes via the Discord API, if Discord integration is enabled and the user has set their Discord username on their profile.
- A new event which is triggered by changing a users Discord username in their profile, which is used for pushing all the user's current roles to Discord.
- A new composer managed dependency on RestCord, which is a PHP Discord API client.

Discord role names match the role display names for member status, team name and tool name, but a mapping table may be needed later.

Two environment variables have been added `DISCORD_TOKEN` and `DISCORD_GUILD_ID`. If these are unset, the event listeners will not subscribe to the events. I can provide values for these environment variables to you on Slack. I think Andrew can invite you to Discord if you want to do some testing.

Other stuff can come later, such as:

- Discord message notifications for things like the membership notifications.
- Removal of Discord roles if a user clears their discordUserId field.
- Discord channel field, like the current Slack channel field on Teams.

Thanks

### CODE REVIEW

**Comment by AaronJackson** [comment]
Perfect timing... Discord is removing the username + discriminator model, so when this happens we'll need to remove the regex user id validation, and the username matching.

Username matching:

https://github.com/NottingHack/hms2/pull/567/files#diff-1bea67ac250db50a8bc0c86b98b7f96729bed10b59095d807eabedf895a6ff12R97-R110

Username validation:

https://github.com/NottingHack/hms2/pull/567/files#diff-8e91484f56a38d906339650a2e3af6dbda84b6a43e6d889b792bd55b53a54e76R141-R144

File: app/Listeners/RoleUpdateDiscordUpdater.php [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+
+    /**
+     * Handles user setting their Discord username field
+     * i.e. push all current roles.
+     *
+     * @param DiscordUsernameUpdated $event
+     */
+    public function onDiscordUsernameUpdated(DiscordUsernameUpdated $event)
+    {
+        $user = $event->user;
```
Comment (dpslwk) [inline app/Listeners/RoleUpdateDiscordUpdater.php:145]: this should likely fetch fresh copies in the same way the above handlers do

**Review by dpslwk — CHANGES_REQUESTED** [review]
Bunch of change requests

How are guild roles added/created in Discord ?
when a new team is added should there be a listener for `RoleCreated` (filtered for `team.*` roles) to auto add them to the server, since then next thing 

`onDiscordUsernameUpdated` does not remove the roes from the old username, should it? the every would need to be passed the original value

I think `Role::MEMBER_TEMPORARYBANNED` and `Role:MEMBER_BANNED` should be explicitly skipped in `onUserAddedToRole` and `onDiscordUsernameUpdated` (even if they don't exist in discord (some idiot might create them)
In fact thinking about it I think the only `member.*` roles that should be allowed are `Role::MEMBER_CURRENT` and `Role::MEMBER_YOUNG`
so there should be filter to only every process `Role::MEMBER_CURRENT, `Role::MEMBER_YOUNG` and any `team.*`

not sue if it's possible but it might have been better to send the dm via a notification and `laravel-notification-channels/discord` https://laravel-notification-channels.com/discord/#contents
going to have to add it anyway to deal with all the notifications that have slack channels, since they will need to be duplicated to discord channels

likely not yet and issue but does `restcord` deal with rate limiting?
I know from another project that the discord notification channel package does not, and its a real PITA

**Comment by AaronJackson** [comment]
Thanks for finding the time to review

Todo:

- [ ] Remove roles when discordUserId is unset
- [ ] Restrict Discord roles to `team.*`, `role.MEMBER_CURRENT` and `role.MEMBER_YOUNG`

> Bunch of change requests
> 
> How are guild roles added/created in Discord ? when a new team is added should there be a listener for `RoleCreated` (filtered for `team.*` roles) to auto add them to the server, since then next thing

I think for the time being Andrew discussed keeping these managed manually. I think for the time being this is reasonable, since the configuration for teams with private channels is also manual. It's not something which changes too often so I think this is ok personally.

> `onDiscordUsernameUpdated` does not remove the roes from the old username, should it? the every would need to be passed the original value

Yeah I need to add this, was sort of putting it off a bit. I think I need to modify the event to include the old username like you've done with the `UserEmailChanged` event.

> I think `Role::MEMBER_TEMPORARYBANNED` and `Role:MEMBER_BANNED` should be explicitly skipped in `onUserAddedToRole` and `onDiscordUsernameUpdated` (even if they don't exist in discord (some idiot might create them) In fact thinking about it I think the only `member.*` roles that should be allowed are `Role::MEMBER_CURRENT` and `Role::MEMBER_YOUNG` so there should be filter to only every process `Role::MEMBER_CURRENT, `Role::MEMBER_YOUNG`and any`team.*`

Will do!

> not sue if it's possible but it might have been better to send the dm via a notification and `laravel-notification-channels/discord` https://laravel-notification-channels.com/discord/#contents going to have to add it anyway to deal with all the notifications that have slack channels, since they will need to be duplicated to discord channels

I was looking at that package today, was planning to do the notifications as a separate branch/PR to avoid a sudden mass in one review. Now you have reviewed though, I'll add it to this PR after fixing the other stuff.

> likely not yet and issue but does `restcord` deal with rate limiting? I know from another project that the discord notification channel package does not, and its a real PITA

No idea to be honest, I'm hoping we won't hit any rate limiting given it's currently only updating roles as and when users update their discordUserId field, or when a user is added to a team. Could be an issue if they were in all the teams though...

Cheers - will look through your other bits below and make those changes.

File: database/migrations_doctrine/Version20230425032647_add_discord_username.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+class Version20230425032647 extends AbstractMigration
+{
+    /**
+     * @param Schema $schema
+     */
+    public function up(Schema $schema): void
+    {
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() != 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('ALTER TABLE profile CHANGE discord_user_id discord_id VARCHAR(32) DEFAULT NULL');
```
[the remark this answers is not in the document]
Comment (AaronJackson) [reply database/migrations_doctrine/Version20230425032647_add_discord_username.php:17]: Thanks, wasn't sure on the process tbh

**Comment by AaronJackson** [comment]
Right - not quite sure if I did the notification stuff right, but it works... 

Made the other changes too

File: app/Notifications/Banking/AuditResult.php [hunk 3]
```diff
@@ -72,7 +75,7 @@ public function __construct(
      */
     public function via($notifiable)
     {
-        return ['mail', 'slack'];
+        return ['mail', 'slack', DiscordChannel::class];
```
Comment (dpslwk) [inline app/Notifications/Banking/AuditResult.php:78]: IIRC Discord notification channel does not handle missing token well

need to check again but having commented out or empty  `DISCORD_TOKEN` in .env might results in the following
`Unresolvable dependency resolving [Parameter #1 [ <required> $token ]] in class NotificationChannels\Discord\Discord `

if this is an issue then `DiscordChannel::class` need to be dynamically added to this list

Comment (dpslwk) [reply app/Notifications/Banking/AuditResult.php:78]: this would applies to all notifications

File: app/HMS/Entities/Profile.php [hunk 4]
```diff
@@ -70,6 +70,11 @@ class Profile
      */
     protected $dateOfBirth;
 
+    /**
+     * @var null|string
+     */
+    protected $discordUserId;
```
Comment (dpslwk) [inline app/HMS/Entities/Profile.php:76]: this is actually the username+discriminator right? not the user_id snowflake

https://discord.com/developers/docs/resources/user#user-object-user-structure

To consider: 
should we change name from `discordUserId ` to `discordUsername` ? 
id say it should be `discordUsernameDiscriminator`  but there changing the soon anyway

**Review by dpslwk — CHANGES_REQUESTED** [review]

File: app/Notifications/Banking/AuditResult.php (hunk 3 above)
Comment (AaronJackson) [reply app/Notifications/Banking/AuditResult.php:78]: yes you're right - thanks for spotting

**Comment by AaronJackson** [comment]
Not all channels have a private channel, but for things like membership this can be useful when discussing responses to emails etc. I've added two fields, `discordChannel` and `discordPrivateChannel`. Only `discordChannel` is displayed to non-admin users.

For the time being, I've set Role routeNotificationForDiscord to use the private channel first, and then fallback to the public one if the private channel isn't set. Might need to modify this at some point if we chose to send other notifications (e.g. replacing google forms with hms)

I decided not to store the snowflake and instead keep the channel names as strings, although I appreciate we might need to change this in future to avoid hitting rate limits if we are sending a large number of notifications. I'm happy to take a look at this eventually, although I'd probably take the approach of adding caching support (via memcached) to the discord helper class, which is doing the name to snowflake resolution

File: app/Notifications/Banking/AuditResult.php [hunk 5]
```diff
@@ -72,7 +75,12 @@ public function __construct(
      */
     public function via($notifiable)
     {
-        return ['mail', 'slack'];
+        $channels = ['mail', 'slack'];
+        if (config('services.discord.token')) {
+            array_push($channels, DiscordChannel::class);
```
Comment (dpslwk) [inline app/Notifications/Banking/AuditResult.php:80]: 🤢old syntax 

```suggestion
            $channels[] = DiscordChannel::class;
```

File: resources/views/role/edit.blade.php [hunk 6]
```diff
@@ -53,6 +53,26 @@
       @endif
     </div>
 
+    <div class="form-group">
+      <label for="discordChannel" class="form-label">Discord channel</label>
+      <input class="form-control" id="discordChannel" type="text" name="discordChannel" value="{{ old('discordChannel', $role->getDiscordChannel()) }}">
+      @if ($errors->has('discordChannel'))
+	<p class="help-text">
```
Comment (dpslwk) [inline resources/views/role/edit.blade.php:60]: indentation? this should just be at the same level

File: resources/views/role/show.blade.php [hunk 7]
```diff
@@ -25,6 +25,16 @@
           <th>Slack Channel:</th>
           <td>{{ $role->getSlackChannel() }}</td>
         </tr>
+        <tr>
+          <th>Discord Channel:</th>
+          <td>{{ $role->getDiscordChannel() }}</td>
+        </tr>
+	@can('role.edit.all')
```
Comment (dpslwk) [inline resources/views/role/show.blade.php:32]: indentation again

File: resources/views/user/edit.blade.php [hunk 8]
```diff
@@ -133,6 +133,21 @@
...
+        <hr>
+
+        <div class="form-group">
+          <label for="discordUsername" class="form-label">Discord Username</label>
+          <input class="form-control" id="discordUsername" type="text" name="discordUsername" value="{{ old('discordUsername', $user->getProfile()->getDiscordUsername()) }}">
+	  <small class="form-text text-muted">
+	    This can be copied by clicking on your username at the bottom left from the Discord app.
+	  </small>
+          @if ($errors->has('discordUsername'))
+            <p class="help-text">
```
Comment (dpslwk) [inline resources/views/user/edit.blade.php:146]: indentation

File: resources/views/role/edit.blade.php [hunk 9]
```diff
@@ -53,6 +53,26 @@
...
+          <strong>{{ $errors->first('discordChannel') }}</strong>
+	</p>
+      @endif
+    </div>
+
+    <div class="form-group">
+      <label for="discordPrivateChannel" class="form-label">Discord channel (Private)</label>
+      <input class="form-control" id="discordPrivateChannel" type="text" name="discordPrivateChannel" value="{{ old('discordPrivateChannel', $role->getDiscordPrivateChannel()) }}">
+      @if ($errors->has('discordPrivateChannel'))
+	<p class="help-text">
```
Comment (dpslwk) [inline resources/views/role/edit.blade.php:70]: indentation

**Review by dpslwk — CHANGES_REQUESTED** [review]
sorry last few nit picks ;)

**Comment by dpslwk** [comment]
we might need to wrap Discord and Slack things in Features flags,
at least the font end/view bits (form fields / display) 

Nottinghack is not the only deployment of HMS and not all deployments will have either of these chat systems

though I guess mostly I'm thinking about not showing this on the teams.index and teams.show page

File: app/Notifications/Banking/AuditResult.php (hunk 5 above)
Comment (AaronJackson) [reply app/Notifications/Banking/AuditResult.php]: :D what happens when you haven't done much php since 2010

Comment (dpslwk) [reply app/Notifications/Banking/AuditResult.php]: I've pretty much been a full time Laravel developer since the start of lock down

**Review by dpslwk — CHANGES_REQUESTED** [review]

**Review by dpslwk — APPROVED** [review]

**Pull request merged**